Fix indexes dont remove previous attribute values on update

phymbert opened the pull request:
Hello,

Indexes still return previous values when attibute changed after updating/remove the same entity.
It also results that indexes still growing and growing if we call update method lot of time with the same entity.
I dont know what is the best way to handle this, I submit a unit test that reflects my problem and a simple fix. Sure you will suggest a better way.

Appreciate your feedback.

phymbert commented on code/src/main/java/com/googlecode/cqengine/index/support/AbstractMapBasedAttributeIndex.java:
```diff
@@ -108,6 +112,21 @@ public boolean removeAll(ObjectSet<O> objectSet, QueryOptions queryOptions) {
             boolean modified = false;
             ConcurrentMap<A, StoredResultSet<O>> indexMap = this.indexMap;
             for (O object : objectSet) {
+              DeduplicationOption deduplicationOption = queryOptions.get(DeduplicationOption.class);
```
Maybe we need another Option type, this is just to show the concept

npgall commented:
I'm away from my computer at the moment but I don't think you are using the update() method correctly. The first argument should be an object you want to remove, and the second argument an object you want to insert. AFAICS you are supplying the same object for both?

phymbert commented:
Thank you for feedback. Actually yes, and it works fine this way with ConcurrentIndexedCollection. Javadoc said it is not allowed only for Transactional implementation.
Anyway this is not the root cause as far as update method just calls sequentially addAll then removeAll. Calling removeAll directly will result with same behavior. Would you please double check ?

npgall commented:
Hi,

This test is invalid. You are creating a new car with the same carId as an existing car in the collection, and then trying to remove the new car from the collection instead of the existing car.

If you replace this:
```
Car carUpdated = new Car(0, "Renault", "Megane", Car.Color.RED, 5, 5000.00, null);
cars.update(Arrays.asList(carUpdated), Arrays.asList(carUpdated),
        queryOptions(deduplicate(DeduplicationStrategy.MATERIALIZE)));
```
With this:
```
Car existingCar = CarFactory.createCar(0);
Car carUpdated = new Car(0, "Renault", "Megane", Car.Color.RED, 5, 5000.00, null);
cars.update(Arrays.asList(existingCar), Arrays.asList(carUpdated),
    queryOptions(deduplicate(DeduplicationStrategy.MATERIALIZE)));
```
...then the test will pass.

phymbert commented:
I do not agree. Let's assume the following:

1. You load an entity from JPA, detach it and put it in the collection
2. The database changed, and you want to update your collection
3. The only way is to call update (or removeAll + addAll) with the same collection, that what the test try to reflect.

As I understand your suggestion is to first query for the old object in the collection and remove it. This is really cost expensive.

npgall commented:
In this case you are completely mis-using the API. The JavaDoc for the update() method explicitly says the first argument is the object(s) to remove from the collection. You didn't supply an object that was already in the collection.

The fact that some indexes were updated and some were not, is because you furthermore gave the same primary key to the new object as one that was already stored in the collection, while at the same time the implementation of equals() relied on this primary key to determine equality. It's a classic mis-use of collection APIs. You need to read more about how collections which rely on hashCode(), compareTo() and equals()/object equality are implemented.

This does relate however, to a legitimate question about how best to update the _fields_ of objects which are already stored in a collection (whether CQEngine or a JDK collection). You can find some background info on this in issue #91.

phymbert commented:
Thank you for feedback, appreciate for your time.

Javadoc states:

> objectsToRemove The objects to remove from the collection

It is not stated: "Objects instances at the state when they were added to the collection (not the identity one specified by equals or hashcode) should be provided - or hashCode/equals are not used".

The issue is giving the same id to removeAll, doesnt remove from indexes if attributes values have changed, despite your explain it clearly here [#91](https://github.com/npgall/cqengine/issues/91), I think it is a lack of the API when updating the collection from database entities and it can be solve easily, specially when you are speaking about primary key or identifier within the API.

The suggested fix allow to handle this case faster, see time results below:
`
  @Test
  public void testUpdateIndexWithFetchFirst() {
    IndexedCollection<Car> cars = new ConcurrentIndexedCollection<Car>();
    cars.addIndex(NavigableIndex.onAttribute(Car.CAR_ID));
    cars.addIndex(NavigableIndex.onAttribute(Car.MANUFACTURER));
    cars.addAll(CarFactory.createCollectionOfCars(100));

    long startTime = System.nanoTime();
    ResultSet<Car> previousCar = cars.retrieve(in(Car.CAR_ID, 0, 1));
    List<Car> removedCars = new ArrayList<Car>();
    for (Car car : previousCar) {
      removedCars.add(car);
    }
    long preQueryTime = System.nanoTime();
    List<Car> updated = new ArrayList<Car>();
    updated.add(new Car(0, "Renault", "Megane", Car.Color.RED, 5, 5000.00, null));
    updated.add(new Car(1, "Citroen", "C4", Car.Color.RED, 5, 5000.00, null));
    cars.update(removedCars, updated);
    long updateTime = System.nanoTime();

    ResultSet<Car> results = cars.retrieve(in(Car.CAR_ID, 0, 1),
        queryOptions(
        orderBy(ascending(Car.CAR_ID)))
        );
    for (Car car : results) {
      switch (car.getCarId()) {
        case 0:
          assertEquals(0, car.getCarId());
          assertEquals("Renault", car.getManufacturer());
          assertEquals("Megane", car.getModel());
          break;
        case 1:
          assertEquals(1, car.getCarId());
          assertEquals("Citroen", car.getManufacturer());
          assertEquals("C4", car.getModel());
          break;
        default:
          fail("Not supported");
          break;
      }
    }
    assertEquals(2, results.size());
    long endTime = System.nanoTime();

    System.out.println(String.format("testUpdateIndexWithFetchFirst time = %fms, preQueryTime=%fms, updateTime=%fms, queryTime=%fms",
        (endTime - startTime) / 1000000f,
        (preQueryTime - startTime )/ 1000000f,
        (updateTime - preQueryTime) / 1000000f,
        (endTime - updateTime) / 1000000f));   
  }

  @Test
  public void testUpdateIndexOption() {
    IndexedCollection<Car> cars = new ConcurrentIndexedCollection<Car>();
    cars.addIndex(NavigableIndex.onAttribute(Car.CAR_ID));
    cars.addIndex(NavigableIndex.onAttribute(Car.MANUFACTURER));
    cars.addAll(CarFactory.createCollectionOfCars(100));

    List<Car> updated = new ArrayList<Car>();
    updated.add(new Car(0, "Renault", "Megane", Car.Color.RED, 5, 5000.00, null));
    updated.add(new Car(1, "Citroen", "C4", Car.Color.RED, 5, 5000.00, null));
    long startTime = System.nanoTime();
    cars.update(updated, updated,
        queryOptions(deduplicate(DeduplicationStrategy.MATERIALIZE)));
    long updateTime = System.nanoTime();

    ResultSet<Car> results = cars.retrieve(in(Car.CAR_ID, 0, 1),
        queryOptions(
        orderBy(ascending(Car.CAR_ID)))
        );
    for (Car car : results) {
      switch (car.getCarId()) {
        case 0:
          assertEquals(0, car.getCarId());
          assertEquals("Renault", car.getManufacturer());
          assertEquals("Megane", car.getModel());
          break;
        case 1:
          assertEquals(1, car.getCarId());
          assertEquals("Citroen", car.getManufacturer());
          assertEquals("C4", car.getModel());
          break;
        default:
          fail("Not supported");
          break;
      }
    }
    assertEquals(2, results.size());
    long endTime = System.nanoTime();
    
    System.out.println(String.format("testUpdateIndexOption time = %fms, preQueryTime=%fms, updateTime=%fms, queryTime=%fms",
        (endTime - startTime) / 1000000f,
        0f,
        (updateTime - startTime) / 1000000f,
        (endTime - updateTime) / 1000000f));    
  }`

Both work but time is really different:

> testUpdateIndexWithFetchFirst time = 2.935007ms, preQueryTime=2.638181ms, updateTime=0.159703ms, queryTime=0.137123ms
testUpdateIndexOption time = 1.115048ms, preQueryTime=0.000000ms, updateTime=0.998452ms, queryTime=0.116596ms

I dont see the issue with the patch, because it can be enabled/disabled on demand. Do you suggest something better ?

npgall commented:
I doubt retrieval takes >2ms for 100 objects. I'd say the latency you measure is the initial lazy class loading at application startup. You might need to run the test multiple times (10K times for hotspot warmup actually).

Remember that indexes are a one-way lookup, and if we don't know the previous value of an attribute, then we cannot update the index efficiently when the attribute value is changed.

Consider an instance of an object which had a field x=2, which was changed to x=3. 
Say you already had an index on field x, which therefore had a mapping 2->{your object}.
The entry for this object needs to be changed from 2->{your object}, to 3->{your object}.

Given an instance of the object whose field x is now 3, how could we locate the old entry in the index? The answer is, as the original value of the field has been lost, you cannot locate the object in the index quickly anymore. So you'd have to scan the entire index, examining every object, just to locate the one entry you need to remove. Such an operation has O(n) time complexity, and I think your patch uses this approach.

If you need to refresh the objects stored in an IndexedCollection whenever rows in a database table have changed, I'd recommend:
(1) When you initially create or subsequently update the IndexedCollection, store a collectionUpdatedTimestamp in a variable somewhere.
(2) Store a lastModifiedTimestamp in each row in your database table.
(3) Periodically SELECT rows from the database table WHERE lastModifiedTimestamp > collectionUpdatedTimestamp, as objectsToAdd.
(4) For each of those objects, determine their primary key, and retrieve the existing versions of those objects from the IndexedCollection into a Set of objectsToRemove.
(5) Call IndexedCollection.update(objectsToRemove, objectsToAdd).

This approach will outperform the approach in your patch, as it has O(1) time complexity for each object to be removed.

phymbert commented:
Agreed, thank you for this nice answer!
